cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/SparkPipelineRunner.java

```diff
@@ -177,168 +177,29 @@ public void runPipeline(PhaseSpec phaseSpec, String sourcePluginType,
       .map(StageSpec::getName)
       .collect(Collectors.toSet());
 
+    processDag(phaseSpec, sourcePluginType, sec, stagePartitions, pluginContext, collectors, pipelinePhase,
+               functionCacheFactory, macroEvaluator, groupedDag, groups, branchers, shufflers);
+  }
+
+  protected void processDag(PhaseSpec phaseSpec, String sourcePluginType, JavaSparkExecutionContext sec,
+                            Map<String, Integer> stagePartitions, PluginContext pluginContext,
+                            Map<String, StageStatisticsCollector> collectors, PipelinePhase pipelinePhase,
+                            FunctionCache.Factory functionCacheFactory, MacroEvaluator macroEvaluator,
+                            CombinerDag groupedDag, Map<String, Set<String>> groups, Set<String> branchers,
+                            Set<String> shufflers) throws Exception {
     Collection<Runnable> sinkRunnables = new ArrayList<>();
+    Map<String, EmittedRecords> emittedRecords = new HashMap<>();
+    //Emitted records and sinkRunnables will be populated as each stage is processed
     for (String stageName : groupedDag.getTopologicalOrder()) {
-      if (groups.containsKey(stageName)) {
```

Comment:
it looks like we're just moving code around so it can be re-used, but it's hard to tell for sure. If there are any logical changes, please comment on those sections.